### DISCUSSION TITLE
Problem creating an index in migration

### DISCUSSION THREAD

**Discussion by dave42w** [SeaQL/sea-orm#1370] [Q&A] [answered]
Hi,
I want to add an index to this table that migration is creating ok

    use sea_orm_migration::prelude::*;
    #[derive(DeriveMigrationName)]
    pub struct Migration;

    #[derive(Iden)]
    pub enum Place {
        Table,
        Id,
        Name,
    }

    #[async_trait::async_trait]
    impl MigrationTrait for Migration {
        async fn up(&self, manager: &SchemaManager) -> Result<(), DbErr> {

            manager
                .create_table(
                    Table::create()
                        .table(Place::Table)
                        .if_not_exists()
                        .col(
                            ColumnDef::new(Place::Id)
                                .integer()
                                .not_null()
                                .auto_increment()
                                .primary_key(),
                        )
                        .col(ColumnDef::new(Place::Name).string().not_null())
                        .to_owned(),
                )
                .await
        }

        async fn down(&self, manager: &SchemaManager) -> Result<(), DbErr> {
            // Replace the sample below with your own migration scripts
 
            manager
                .drop_table(Table::drop().table(Place::Table).to_owned())
                .await
        }
    }

I have been looking at https://www.sea-ql.org/SeaORM/docs/next/migration/writing-migration/ but can't find an example of a migration creating an index.

I have created a separate migration file for the index but wonder if I could create the index at the same time as the table in a single migration file?

This is the actual code I am trying to create and drop the index:

    use sea_orm_migration::prelude::*;

    #[derive(DeriveMigrationName)]
    pub struct Migration;

    #[derive(Iden)]
    pub enum Place {
        Table,
        Id,
        Name,
    }
    #[async_trait::async_trait]
    impl MigrationTrait for Migration {
        async fn up(&self, manager: &SchemaManager) -> Result<(), DbErr> {

            manager
                .create_index(
                    Index::create()
                        .name("idx-place-name")
                        .unique()
                        .table(Place::Table)
                        .col(Place::Name)                        
                        .col(Place::Id)                        
                        .to_owned(),
                )
                .await
        }

        async fn down(&self, manager: &SchemaManager) -> Result<(), DbErr> {
            // Replace the sample below with your own migration scripts

            manager
                .drop_index(Index::drop().name("idx-place_name").to_owned())
                .await
        }
    }

But when I cargo run the migration script I get

    Compiling migration v0.1.0 (/home/dave/Code/Rust/daves_dashboard/migration)
        Finished dev [unoptimized + debuginfo] target(s) in 12.98s
         Running `/home/dave/Code/Rust/daves_dashboard/target/debug/migration`
    Applying all pending migrations
    Applying migration 'm20230101_000001_create_index'
    Execution Error: error returned from database: (code: 1) no such table: main.place

I don't understand. Where is "main.place" coming from? 

My main application calls the migratory from main.rs

    #[tokio::main]
    async fn main() {
        dotenv().expect(".env file not found");

        let template_dir = env::var("TEMPLATES").unwrap();
        let db_uri: String = env::var("DATABASE_URL").unwrap();
        let d: DatabaseConnection = Database::connect(db_uri).await.unwrap();

        Migrator::up(&d, None).await.unwrap();

        let state = AppState {
            templates: get_templates(template_dir),
            db: d,

        };

        run(state).await
    }

When I run that I get:

     Running `target/debug/daves_dashboard`
    thread 'main' panicked at 'called `Result::unwrap()` on an `Err` value: Exec(SqlxError(Database(SqliteError { code: 1, message: "no   such table: main.place" })))', src/main.rs:44:34
    note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
    [Finished running. Exit status: 101]
    
From a debugging perspective is there a way to see the SQL output from the migration so that I can see what is happening?

Thanks

**Comment by dave42w** [comment]
I think I have sorted this out myself. This appears to be working as I want it (it probably isn't very elegant or idiomatic Rust). I'd love to see an example of this in the docs in the future. If I knew this was the appropriate solution I'd be happy to add to the docs.

    use sea_orm_migration::prelude::*;

    #[derive(DeriveMigrationName)]
    pub struct Migration;

    #[derive(Iden)]
    pub enum Place {
        Table,
        Id,
        Name,
    }

    #[async_trait::async_trait]
    impl MigrationTrait for Migration {
        async fn up(&self, manager: &SchemaManager) -> Result<(), DbErr> {
            let r = manager
            .create_table(
                Table::create()
                .table(Place::Table)
                .if_not_exists()
                .col(
                    ColumnDef::new(Place::Id)
                    .integer()
                    .not_null()
                    .auto_increment()
                    .primary_key(),
                )
                .col(ColumnDef::new(Place::Name).string().not_null())
                .to_owned(),
            )
            .await;
            
            match r {
                Ok(_) => {
                    let r2 = manager
                    .create_index(
                        Index::create()
                        .name("idx-place-name")
                        .unique()
                        .table(Place::Table)
                        .col(Place::Name)                        
                        .col(Place::Id)                        
                        .to_owned(),
                    )
                    .await;
                    return r2;  
                }
                Err(_) => return r,
            }
        }
        
        async fn down(&self, manager: &SchemaManager) -> Result<(), DbErr> {
            // Replace the sample below with your own migration scripts           
            let r = manager
            .drop_index(Index::drop().name("idx-place-name").to_owned())
            .await;

            match r {
                Ok(_) => {
                    let r2 = manager
                    .drop_table(Table::drop().table(Place::Table).to_owned())
                    .await;
                    return r2;
                }
                Err(_) => return r,
            }
        }
    }

**Comment by billy1624** [comment, marked as answer]
Hey @dave42w, I think we can simply use `?` to cancel the migration process when we encounter any errors alone the way. Please check below:

> A documentation PR would be welcomed :)

```rs
use sea_orm_migration::prelude::*;

#[derive(DeriveMigrationName)]
pub struct Migration;

#[derive(Iden)]
pub enum Place {
    Table,
    Id,
    Name,
}

#[async_trait::async_trait]
impl MigrationTrait for Migration {
    async fn up(&self, manager: &SchemaManager) -> Result<(), DbErr> {
        manager
            .create_table(
                Table::create()
                .table(Place::Table)
                .if_not_exists()
                .col(
                    ColumnDef::new(Place::Id)
                    .integer()
                    .not_null()
                    .auto_increment()
                    .primary_key(),
                )
                .col(ColumnDef::new(Place::Name).string().not_null())
                .to_owned(),
            )
            .await?; // Unwrap the error here, cancel the migration on error
        
        manager
            .create_index(
                Index::create()
                .name("idx-place-name")
                .unique()
                .table(Place::Table)
                .col(Place::Name)                        
                .col(Place::Id)                        
                .to_owned(),
            )
            .await?;  // Unwrap the error here, cancel the migration on error

        Ok(()) // All good!
    }
    
    async fn down(&self, manager: &SchemaManager) -> Result<(), DbErr> {
        manager
            .drop_index(Index::drop().name("idx-place-name").to_owned())
            .await?;

        manager
            .drop_table(Table::drop().table(Place::Table).to_owned())
            .await?;

        Ok(())
    }
}
```

**Reply by dave42w** [reply]
Thanks, much neater.

So this would cause a panic if the migration failed, seems apprpriate.
